File: 720-词典中最长的单词.py
from typing import List


class Solution:
    def longestWord(self, words: List[str]) -> str:
        res = ""
        hash = {""}
        # str不能取负, 因此先长度降序，字典升序，再反转
        words.sort(key=lambda x: (-len(x), x), reverse=True)
        for word in words:
            if word[:-1] in hash:
                res = word
                hash.add(word)
        return res

File: 720-词典中最长的单词.cpp
#include <iostream>
using namespace std;
#include <vector>
#include <string>
#include <unordered_set>
#include <algorithm>

class Solution
{
public:
    string longestWord(vector<string> &words)
    {
        string res = "";
        unordered_set<string> hash;
        hash.emplace(res);
        sort(words.begin(), words.end(), [](const string &a, const string &b)
             {
                 if (a.size() != b.size())
                 {
                     return a.size() < b.size(); // 长度升序
                 }
                 else
                 {
                     return a > b; // 字典降序
                 }
             });
        for (auto &word : words)
        {
            if (hash.count(word.substr(0, word.size() - 1)))
            {
                res = word;
                hash.emplace(word);
            }
        }
        return res;
    }
};